### PR TITLE
Add DataReplay to drive the prognostic iterator with a DataSource

### LINKED ISSUE #984: 🚀[FEA]: Add DataReplay to drive the prognostic iterator with a DataSource
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

## Problem

There is currently no built-in way to drive Earth2Studio's prognostic-iterator pipelines (`run.ensemble`, `run.deterministic`, `create_iterator`) with reanalysis data instead of a forecast rollout. Users who want to:

- provide a reference trajectory (e.g. replay ERA5) to score forecasts against,
- feed a downstream prognostic with analysis frames rather than a forecast (for example the coarse input to a temporal interpolator), or
- sub-sample a finer-resolution source in time,

have to hand-roll a loop around `fetch_data`, re-implementing the coordinate handling and the iterator contract each time.

The closest existing pieces don't cover this: `Persistence` echoes the initial condition forward (it never fetches new data), and `ForecastSource` is a data abstraction (init-time × lead-time), not a `PrognosticModel` that plugs into the iterator.

## Proposed solution

A small, weightless `DataReplay` prognostic **adapter** (`torch.nn.Module` + `PrognosticMixin`) that wraps any `DataSource` and steps it through the `create_iterator` interface: the initial condition is the source at `time0`, and each subsequent step fetches the source at the next valid time (advancing `lead_time` by a configurable `step`). It complements `Persistence`, reuses the standard coordinate handshakes, and adds no third-party dependencies.

Implemented in PR #985 .

### PR DESCRIPTION
**Pull request by gertln**
## Description

Add `DataReplay`, a prognostic adapter that steps `earth2studio.data.DataSource` through the `PrognosticModel` iterator interface. Instead of forecasting, each step after the initial condition fetches the source at the next valid time, yielding the source's own reanalysis/analysis frames.

This lets pipelines built around the prognostic iterator be driven by data sources rather than a forecast rollout e.g. supplying a temporal coarse input to a temporal interpolator, providing a reference trajectory to score forecasts against, or sub-sampling a finer source in time. It complements `Persistence` (which echoes the initial state forward); for the forecast-as-trajectory case.

Closes #984 
Blocked by: none (self-contained; no external dependencies or weights).
Related: PR `InterpCRPSDiT` is the first concrete consumer.

### Model details

A `torch.nn.Module` + `PrognosticMixin` adapter with no checkpoint and no forward network. It wraps a `DataSource` (queried by absolute valid time) behind `input_coords` / `output_coords` / `create_iterator`.

### Dependencies added

None

## Checklist

- [x] I am familiar with the Contributing Guidelines.
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes (`models_px.rst`, docstrings).
- [x] The `CHANGELOG.md` is up to date with these changes.
- [x] An issue is linked to this pull request.
- [ ] Assess and address Greptile feedback (AI code review bot).

### CODE REVIEW

**Comment by NickGeneva** [comment]
TODO: Check https://github.com/NVIDIA/earth2studio/blob/main/earth2studio/serve/client/e2client.py#L314

File: earth2studio/models/px/datareplay.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        handshake_dim(input_coords, "lat", -2)
+        handshake_dim(input_coords, "lon", -1)
+        handshake_coords(input_coords, target, "variable")
+        handshake_coords(input_coords, target, "lat")
+        handshake_coords(input_coords, target, "lon")
+        output_coords["batch"] = input_coords["batch"]
+        output_coords["time"] = input_coords.get("time", np.empty(0))
+        output_coords["lead_time"] = (
+            output_coords["lead_time"] + input_coords["lead_time"][-1]
+        )
```
[the remark this answers is not in the document]
Comment (gertln) [reply earth2studio/models/px/datareplay.py]: The intermediate lead_time was actually read in the final line RHS (... + input_coords["lead_time"][-1]). Agreed it reads cleaner inlined. Changed that but kept the behavior.

File: earth2studio/models/px/datareplay.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        x : torch.Tensor
+            Input tensor (used only for batch size / device / dtype).
+        coords : CoordSystem
+            Input coordinate system; must contain ``time``.
+
+        Returns
+        -------
+        tuple[torch.Tensor, CoordSystem]
+            The fetched frame and its coordinate system, one ``step`` ahead.
+        """
```
[the remark this answers is not in the document]
Comment (gertln) [reply earth2studio/models/px/datareplay.py]: Added a Notes section to `__call__` clarifying that front_hook/rear_hook fire only in `create_iterator`. That matches `Persistence`, where hooks are also iterator-scoped by the PrognosticMixin contract. `__call__` is a single advance-step and stays hook-free by design.

**Comment by gertln** [comment]
> TODO: Check https://github.com/NVIDIA/earth2studio/blob/main/earth2studio/serve/client/e2client.py#L314

I had a look and checked the overlap. Both implementations present source data through a prognostic interface but they serve different functions and follow different contracts.

`InferenceOutputModel ` is intended for consuming the finite output of a completed inference workflow. It ignores external initial-condition data, walks the stored valid-time axis until exhausted and can report frames using absolute-time or lead-relative coordinates.

`DataReplay ` follows the usual IC-driven Earth2Studio prognostic-model contract: `create_iterator(x, coords)` accepts an existing initial condition, yields it before advancing and then fetches subsequent source frames at the configured interval. The initial field also establishes the batch size, device, and dtype and can serve as the first endpoint for downstream models such as a temporal interpolator. Accepting only a time and fetching the first frame internally would resemble `InferenceOutputModel ` but would not follow this usual IC-driven interface.

`DataReplay` can also replay stored inference results through `InferenceOutputSource`, if users want those results to follow the usual initial-condition-driven prognostic interface at a chosen time interval.

**Comment by NickGeneva** [comment]
coverage: `earth2studio/models/px/datareplay.py                       71      0   100%`

Excellent work!

**Review by NickGeneva — APPROVED** [review]
